Source/AMRInterpolator/SurfaceExtraction.hpp

```diff
@@ -0,0 +1,165 @@
+    std::vector<integrand_t> m_integrands;
+    std::vector<std::array<IntegrationMethod, 2>> m_integration_methods;
+    std::vector<std::reference_wrapper<std::vector<double>>> m_integrals;
+
+    bool m_done_extraction; //!< whether or not the extract function has been
+                            //!< called for this object
+
+    //! returns the flattened index for m_interp_data and m_interp_coords
+    //! associated to given surface, u and v indices
+    int index(int a_isurface, int a_iu, int a_iv) const
+    {
+        return a_isurface * m_num_points + a_iu * m_params.num_points_v + a_iv;
+    }
+
+  public:
+    //! Normal constructor which requires vars to be added after construction
+    //! using add_var or add_vars
+    SurfaceExtraction(const SurfaceGeometry &a_geom, const params_t &a_params,
+                      double a_dt, double a_time, bool a_first_step,
+                      double a_restart_time = 0.0);
+
+    //! add a single variable or derivative of variable
+    void add_var(int a_var, const Derivative &a_deriv = Derivative::LOCAL);
+
+    //! add a vector of variables/derivatives of variables
+    void add_vars(const std::vector<std::pair<int, Derivative>> &a_vars);
+
+    //! add a vector of variables (no derivatives)
+    void add_vars(const std::vector<int> &a_vars);
+
+    //! Alternative constructor with a predefined vector of variables and
+    //! derivatives
+    SurfaceExtraction(const SurfaceGeometry &a_geom, const params_t &a_params,
+                      const std::vector<std::pair<int, Derivative>> &a_vars,
+                      double a_dt, double a_time, bool a_first_step,
+                      double a_restart_time = 0.0);
+
+    //! Another alternative constructor with a predefined vector of variables
+    //! no derivatives
+    SurfaceExtraction(const SurfaceGeometry &a_geom, const params_t &a_params,
+                      const std::vector<int> &a_vars, double a_dt,
+                      double a_time, bool a_first_step,
+                      double a_restart_time = 0.0);
+
+    //! Do the extraction
+    template <typename InterpAlgo>
+    void extract(AMRInterpolator<InterpAlgo> *a_interpolator);
+
+    //! Add an integrand dependent on the interpolated data over the surface
+    //! for integrate() to integrate over.
+    //! Note the area_element is already included from the SurfaceGeometry
+    //! template class
+    void add_integrand(
+        const integrand_t &a_integrand, std::vector<double> &out_integrals,
+        const IntegrationMethod &a_method_u = IntegrationMethod::trapezium,
+        const IntegrationMethod &a_method_v = IntegrationMethod::trapezium);
+
+    //! Add an integrand which is just a single var. The a_var argument should
+    //! correspond to the order in which the desired var was added to this
+    //! object with add_var
+    void add_var_integrand(
+        int a_var, std::vector<double> &out_integrals,
```

Comment:
I think `a_var` should be be the index in `UserVariables` and not the index in the variables you added with `add_var`. This is easy as you only need to check the variable requested is inside the `m_vars` vector. This was a cause of error in the example I built. If I do `add_var(c_rho)`, I would probably want to follow with `add_var_integrand(c_rho, ...)`.
PLUS. And this is a less trivial re-write of the class: I think it would be a better idea if "labels" were passed each time an integrand is added, and saved inside the class. And not in the end a full vector for `write_integrals`. Why? Well, the biggest reason is so that when I do `àdd_var_integrand` I don't have to specify the label, because from `variable_names` you already have it. Like this you would specify a label for normal integrands as usual and no need to specify it for variable integrands.

Reply:
I don't think these changes are compatible with some `var`s being derivatives.

Reply:
We could automate it by adding a label as "d" + name + "/dx" or something. The cost would be that the user loses liberty to give personalized names. I'm not sure if this is what you meant by compatible.

Reply:
OK, what I meant is that it's not compatible with your first suggestion. I don't like removing the user's choice for `add_var_integrand` but I think choosing the label at the `add_integrand` stage is not a bad idea.

Reply:
Ok. If this ends up being smth to do, we could leave it as a choice for the user, but have a default that gets automatically filled for the case of the `add_var_integrand`

Reply:
Actually, having looked at the code again, I don't want to do this because the `write_integrals` function is separated from the integration and this allows the user to choose which integrals are written (they might just want to use some in the code without writing). I think it will be too difficult to store labels at the `add_integrand` stage and then match them with integrals passed to `write_integrals` so therefore I propose keeping things as they are.

Reply:
This seems reasonable.